[ASTGen] Infrastructure for implementing ASTGen incrementally

rintaro opened the pull request:
* `ASTGenVisitor` has a reference to a legacy C++ Parser configured for ASTGen.
* When `ASTGenVisitor` encounters a AST node that hasn't been migrated, call the C++ `Parser::parse(Decl|Stmt|Expr|Type)` to parse the position using the legacy parser.
* If the legacy parser is configured for ASTGen, it calls ASTGen's  `swift_ASTGen_build(Decl|Stmt|Expr|Type)` for each ASTNode parsing (unless the call is not directly from the ASTGen.)



rdar://117151886

rintaro committed:
[ASTGen] Generalize findSyntaxNodeInSourceFile(wantOutermost:true)

When 'wantOutermost' is true, traverse the parent as long as the
position is the same as the target position.

rintaro committed:
[ASTGen] Infrastructure to implment ASTGen Incrementally

* 'ASTGenVisitor' has a reference to a legacy C++ Parser configured for
  ASTGen.
* If 'ASTGenVisitor' encounters a AST node that hasn't been migrated,
  call parse(Decl|Stmt|Expr|Type) to parse the position using the legacy
  parser.
* The legacy parser calls ASTGen's
  'swift_ASTGen_build(Decl|Stmt|Expr|Type)' for each ASTNode "parsing"
  (unless the call is not directly from the ASTGen.)

rdar://117151886

ahoppen commented on include/swift/Parse/Parser.h:
```diff
@@ -158,6 +158,11 @@ class Parser {
 
+#if SWIFT_BUILD_SWIFT_SYNTAX
+  // This Parser is a fallback parser for ASTGen.
+  bool IsForASTGen = false;
```
Why did you choose to put this behind a `#if`? Seems like a potential source of issues in the future if you use it from somewhere that isn’t guarded by `SWIFT_BUILD_SWIFT_SYNTAX` and that you won’t see locally if you have `SWIFT_BUILD_SWIFT_SYNTAX` enabled.

rintaro replied:
I don't think "we" (developers with `SWIFT_BUILD_SWIFT_SYNTAX ` environment) would do much in `#else` branches in Parser.  I'm more worried that developers for non-`SWIFT_BUILD_SWIFT_SYNTAX` environments use `IsForASTGen` mistakingly.

ahoppen replied:
What do you imagine someone might use `IsForASTGen` for that would be guarded by the `#if`?

rintaro replied:
I have no idea. But if it's inside `#if SWIFT_BUILD_SWIFT_SYNTAX` they cannot use it by mistake. So I thought it's safer to guard it.
Though I'm totally OK not to guard this, if you think that's better.

ahoppen replied:
It’s fine. If you think it provides value, let’s guard it.

hamishknight replied:
FWIW we have had issues in the past where `#if`'ing out fields leads to crashes if the library and its client ever disagree about the condition (IIRC most recently we had issues with `#ifndef NDEBUG`). Not sure if that can be the case here, but I would tend on the side of not `#if`'ing out fields in general.

ahoppen commented on include/swift/Parse/Parser.h:
```diff
@@ -957,7 +962,8 @@ class Parser {
                                bool IfConfigsAreDeclAttrs,
-                               llvm::function_ref<void(Decl *)> Handler);
+                               llvm::function_ref<void(Decl *)> Handler,
+                               bool fromASTGen = false);
```
Could you add a comment what `fromASTGen` does? Similar for the other functions that have `fromASTGen`.

---

Also, all the other parameter start with uppercase.

rintaro replied:
I _think_ we are slowly migrating to "camelCase" for function parameters. I prefer new parameters being lower camel case.

ahoppen replied:
Disclaimer: I don’t know what the transition path is here but at least I have always tried to be consistent with the rest of the surrounding code. But I’ll leave it up to you, ultimately.

ahoppen commented on include/swift/Parse/Parser.h:
```diff
@@ -2075,6 +2039,14 @@ class Parser {
+  //===--------------------------------------------------------------------===//
+  // ASTGen support.
+
+  ParserResult<TypeRepr> parseTypeReprFromSyntaxTree();
```
Just reading the function declaration, I’m thinking: Which syntax tree is this parsing from? I think a doc comment could help explain this (or a need a better name).

ahoppen commented on lib/ASTGen/Sources/ASTGen/Exprs.swift:
```diff
@@ -1,6 +1,57 @@
+      _ where current.is(ExprSyntax.self):
+      return false
+    default:
+      break;
```
```suggestion
      break
```

ahoppen commented on lib/ASTGen/Sources/ASTGen/Exprs.swift:
```diff
@@ -1,6 +1,57 @@
+    if current.id == node.id {
+      return true
+    }
+    current = current.parent!
```
I don’t think `current.parent` needs to be non-nil if the entire syntax tree starts with an `ExprSyntax`. I think we should return `true` if there is no more parent.

rintaro replied:
This is walking up the parents from the first token of `node`. `.parent` must exist if `current` is not `node`

ahoppen replied:
Oh, I see. Could you add a comment that contains exactly what you said in the reply?

ahoppen commented on lib/ASTGen/Sources/ASTGen/Exprs.swift:
```diff
@@ -1,6 +1,57 @@
+        .closureExpr, .functionCallExpr, .declReferenceExpr, .memberAccessExpr,
+        .tupleExpr:
+      break
+    case // Known unimplemented kinds.
+        .arrayExpr, .arrowExpr, .asExpr, .assignmentExpr, .awaitExpr,
+        .binaryOperatorExpr, .booleanLiteralExpr, .borrowExpr, .canImportExpr,
+        .canImportVersionInfo, .dictionaryExpr, .discardAssignmentExpr, .doExpr,
+        .editorPlaceholderExpr, .floatLiteralExpr, .forceUnwrapExpr, .inOutExpr,
+        .infixOperatorExpr, .integerLiteralExpr, .isExpr, .keyPathExpr,
+        .macroExpansionExpr, .consumeExpr, .copyExpr, .nilLiteralExpr,
+        .optionalChainingExpr, .packElementExpr, .packExpansionExpr,
+        .postfixIfConfigExpr, .postfixOperatorExpr, .prefixOperatorExpr,
+        .regexLiteralExpr, .sequenceExpr, .genericSpecializationExpr,
+        .stringLiteralExpr, .simpleStringLiteralExpr, .subscriptCallExpr,
+        .superExpr, .switchExpr, .ternaryExpr, .tryExpr, .typeExpr,
+        .unresolvedAsExpr, .unresolvedIsExpr, .patternExpr,
+        .unresolvedTernaryExpr:
+      return false
```
Is there any value in tracking the known unimplemented kinds? Or is this meant as a To-Do list of nodes to migrate?

rintaro replied:
Yeah, I consider this is a to-do list. If we have a `ExprSyntax` specific `SyntaxKind` or `SyntaxEnum` it should e great.

ahoppen replied:
👍🏽 

FWIW, adding an `ExprSyntaxEnum` to swift-syntax would make total sense to me.

ahoppen commented on lib/ASTGen/Sources/ASTGen/LegacyParse.swift:
```diff
@@ -0,0 +1,43 @@
+import BasicBridging
+import SwiftSyntax
+import ParseBridging
+
+
+
```
Many whitespaces

ahoppen commented on lib/ASTGen/Sources/ASTGen/LegacyParse.swift:
```diff
@@ -0,0 +1,43 @@
+import ASTBridging
```
Shouldn’t we have copyright headers in ASTGen?

ahoppen commented on lib/ASTGen/Sources/ASTGen/LegacyParse.swift:
```diff
@@ -0,0 +1,43 @@
+
+  func generateWithLegacy(_ node: ExprSyntax) -> BridgedExpr {
+    // NOTE: Postfix expressions and sequence expressions share the same start
+    // location with the inner expression. This function must only be called on
+    // "top" expressions. Even if that expression is migrated.
```
I’m not entirely sure what you mean by top expressions here. Maybe an example would help

rintaro replied:
Rephrased.

ahoppen replied:
It’s a lot clearer now. Thanks!

ahoppen commented on lib/ASTGen/Sources/ASTGen/Types.swift:
```diff
@@ -3,6 +3,40 @@ import BasicBridging
+    if current.id == node.id {
+      return true
+    }
+    current = current.parent!
```
Same comment to `parent` being `nil` here as for expressions.

ahoppen commented on lib/DriverTool/swift_parse_test_main.cpp:
```diff
@@ -54,6 +59,9 @@ struct SwiftParseTestOptions {
       llvm::cl::desc("skip function bodies and type members if possible"));
 
+  llvm::cl::opt<bool> Dump = llvm::cl::opt<bool>(
+      "dump", llvm::cl::desc("dump result for each iteration"));
```
What exactly is this supposed to dump? The AST?

rintaro replied:
Whatever the result. for libParse and ASTGen, the result is AST. For SwiftParser, I'm planning to just dump the syntax tree. It's just convenient to confirm that it actually performed the parsing.

ahoppen replied:
Ah, OK 👍🏽

ahoppen commented on lib/ASTGen/Sources/ASTGen/LegacyParse.swift:
```diff
@@ -0,0 +1,43 @@
+extension ASTGenVisitor {
+
+  private func bridgedSourceLocForParse(_ node: some SyntaxProtocol) -> BridgedSourceLoc {
+    return BridgedSourceLoc(at: node.position, in: self.base)
```
I wonder if this should be `positionAfterSkippingLeadingTrivia` because that corresponds to the start location of the token in the C++ AST.

rintaro replied:
It doesn't really matter, but you are right. I removed this method. and changed to use the existing `SyntaxProtocol.bridgedSourceLoc(in: ASTGenVisitor)`  method.

ahoppen commented on lib/Parse/ParseBridging.cpp:
```diff
@@ -0,0 +1,180 @@
+  P.restoreParserPosition(PP);
+  ParserResult<Expr> result =
+      P.parseExprImpl(diag::expected_expr, isExprBasic, /*fromASTGen=*/true);
+  return result.getPtrOrNull();
```
We somehow also need to propagate `hasError` and `hasCodeCompletionToken`. Is that something you want to do in this PR or in a follow-up?

ahoppen commented on lib/Parse/ParseBridging.cpp:
```diff
@@ -0,0 +1,180 @@
+                                           BridgedDeclContext DC,
+                                           bool isExprBasic) const {
+  auto &P = unbridged();
+  auto PP = P.getParserPosition(loc.unbridged(), loc.unbridged());
```
Do you also want to add the `/*PreviousLoc=*/ annotation here? I think it really helps.

ahoppen commented on lib/Parse/ParseBridging.cpp:
```diff
@@ -0,0 +1,180 @@
+  P.CurDeclContext = DC.unbridged();
+  P.restoreParserPosition(PP);
+
+  SmallVector<Decl *, 2> decls;
```
Why do we collect the `decls` here? I don’t think we do anything with them.

rintaro replied:
No reason. Removed.

rintaro commented on lib/ASTGen/Sources/ASTGen/Exprs.swift:
```diff
@@ -1,6 +1,65 @@
+      break
+    case // Known unimplemented kinds.
+        .arrayExpr, .arrowExpr, .asExpr, .assignmentExpr, .awaitExpr,
+        .binaryOperatorExpr, .booleanLiteralExpr, .borrowExpr, .canImportExpr,
```
Literals are already implemented. I didn't notice they were in `Literals.swift`.

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[ASTGen] Remove ASTGenTypes experimental feature

There's no reason to generate only TypeRepr using ASTGen anymore.
Use ParserASTGen feature to test test/ASTGen/types.swift because
ASTGen now can generate the whole test file for type checking.

rintaro committed:
[ASTGen] Use C briding instead of C++ method

Apparently C++ method bridging doesn't work in Linux?

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
preset=asan
@swift-ci Please test with preset macOS platform

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
preset=asan
@swift-ci Please test with preset macOS platform

DougGregor commented:
I like this approach of bouncing between the two parsers as needed, so we can opt in to ASTGen production-by-production.

rintaro committed:
[ASTGen] Disable a test for the same ASAN failure as other ASTGen test

rintaro commented:
@swift-ci Please smoke test

rintaro commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -477,21 +510,155 @@ extension Optional where Wrapped: LazyCollectionProtocol {
-  }
+  else {
+    // FIXME: Produce an error
+    return nil
```
Currently `findSyntaxNodeInSourceFile()` prints an error message to stdout, which is not great, but at least it prevents "unwanted surprises". But yeah, I think we should add a diagnostic on the diagnostic engine.
But since this PR is getting bigger, and to unblock some follow-up patches I have, I'd like to do it in the followups.
Also, I'm thinking to change `findSyntaxNodeInSourceFile()` itself. https://github.com/apple/swift/pull/69724#discussion_r1387258659

rintaro commented:
preset=asan
@swift-ci Please test with preset macOS platform

rintaro commented on test/ASTGen/types.swift:
```diff
@@ -1,7 +1,8 @@
 
 // -enable-experimental-feature requires an asserts build
 // REQUIRES: asserts
+// REQUIRES: rdar116686158
```
It is really unfortunate that this test hits ASAN failure too. We're actively looking into this issue.
Anyway I confirmed these tests passes locally.

hamishknight replied:
Should we mark it `UNSUPPORTED: asan` for now so we at least don't lose the test coverage?

bnbarham commented on lib/DriverTool/swift_parse_test_main.cpp:
```diff
@@ -113,6 +134,67 @@ struct SwiftParserExecutor {
   }
 };
 
+static void appendToVector(void *declPtr, void *vecPtr) {
```
Is this used? Only reference I see is in ParseDecl.cpp

rintaro replied:
Ah, it's a leftover. I'm removing it in a followup

hamishknight reviewed (commented):
Cool!
